git_machete/git_operations.py

```diff
@@ -0,0 +1,412 @@
+    def spoonfeed_log_shas(self, b: str) -> Generator[str, None, None]:
+        if b not in self.initial_log_shas_cached:
+            self.initial_log_shas_cached[b] = self.log_shas(b, max_count=MAX_COUNT_FOR_INITIAL_LOG)
+        for sha in self.initial_log_shas_cached[b]:
+            yield sha
+
+        if b not in self.remaining_log_shas_cached:
+            self.remaining_log_shas_cached[b] = self.log_shas(b, max_count=None)[MAX_COUNT_FOR_INITIAL_LOG:]
+        for sha in self.remaining_log_shas_cached[b]:
+            yield sha
+
+    # Reflog magic
+    def load_all_reflogs(self) -> None:
+        # %gd - reflog selector (refname@{num})
+        # %H - full hash
+        # %gs - reflog subject
+        all_branches = [f"refs/heads/{b}" for b in self.local_branches()] + \
+                       [f"refs/remotes/{self.combined_counterpart_for_fetching_of_branch(b)}" for b in self.local_branches() if self.combined_counterpart_for_fetching_of_branch(b)]
+        # The trailing '--' is necessary to avoid ambiguity in case there is a file called just exactly like one of the branches.
+        entries = utils.non_empty_lines(self.popen_git("reflog", "show", "--format=%gD\t%H\t%gs", *(all_branches + ["--"])))
+        self.reflogs_cached = {}
+        for entry in entries:
+            values = entry.split("\t")
+            if len(values) != 3:  # invalid, shouldn't happen
+                continue
+            selector, sha, subject = values
+            branch_and_pos = selector.split("@")
+            if len(branch_and_pos) != 2:  # invalid, shouldn't happen
+                continue
+            b, pos = branch_and_pos
+            if b not in self.reflogs_cached:
+                self.reflogs_cached[b] = []
+            self.reflogs_cached[b] += [(sha, subject)]
+
+    def reflog(self, b: str) -> List[REFLOG_ENTRY]:
+        # git version 2.14.2 fixed a bug that caused fetching reflog of more than
+        # one branch at the same time unreliable in certain cases
+        if self.get_git_version() >= (2, 14, 2):
+            if self.reflogs_cached is None:
+                self.load_all_reflogs()
+            return self.reflogs_cached.get(b, [])
+        else:
+            if self.reflogs_cached is None:
+                self.reflogs_cached = {}
+            if b not in self.reflogs_cached:
+                # %H - full hash
+                # %gs - reflog subject
+                self.reflogs_cached[b] = [
+                    tuple(entry.split(":", 1)) for entry in utils.non_empty_lines(  # type: ignore
+                        # The trailing '--' is necessary to avoid ambiguity in case there is a file called just exactly like the branch 'b'.
+                        self.popen_git("reflog", "show", "--format=%H:%gs", b, "--"))
+                ]
+            return self.reflogs_cached[b]
+
+    def create_branch(self, b: str, out_of_revision: str) -> None:  # maybe move such a methods into github.py
```

Comment:
Hmmm why github.py? 🤔 this method is independent from GitHub, it still makes sense wherever the remote repository is hosted... or even when there is no remote repo configured at all

Reply:
The only methods that need to be kept under github.py are the ones that strictly talk to GitHub API, handling things that are **not** known to `git` as such &mdash; like pull requests.

Reply:
Hmm, as every method in git_operations belongs now to GitContext, this comment is outdated :)